### ISSUE TITLE
type narrowing not working inside array forEach callback

### ISSUE THREAD

**Issue by joshribakoff** [microsoft/TypeScript#40202]
**TypeScript Version:**  3.9.7
**Search Terms:**

type narrowing not working inside array forEach callback

**Code**

```ts
type Observer = {next:()=>void}

let observer: Observer|undefined

const value = {}

if (observer) {
  Object.keys(value).forEach((name) => {
    observer.next();
  });
}
```

**Expected behavior:** No error, narrows type

**Actual behavior:** Error, doesn't narrow type

**Playground Link:** https://www.typescriptlang.org/play?#code/C4TwDgpgBA8gRgZwgJwG4qgXigbwHYQAewAXABQCUmAfKgPYCWAJgL4BQbANhMFHYinTISsAWhQAfAK54mEAGYMCTDgGM6eBL1QBDTlOjYc7Ng3lQy-JOOQVcbKKIBWEVcAB0AawggEZXfoQFO7ydMgAojqqABZkZHg6ALZBWNT2jo5WgijuBMSUANwOUCwURSxAA

**Related Issues:** 
https://github.com/microsoft/TypeScript/issues/36436
https://github.com/microsoft/TypeScript/issues/9998

**Work arounds**

- enumerate the array even if i don't want to do any work (downside is its not performant)
- check if `observer` is defined multiple times (downside is TS imposing on code readability)
```ts
  if (observer) {
    Object.keys(value).forEach((name) => {
      if (observer) {
        observer.next();
      }
    });
  }
```

**Comment by MartinJohns** [comment]
Duplicate of #9998.
